c2rust-transpile/src/c_ast/conversion.rs

```diff
@@ -1934,12 +1974,121 @@ impl ConversionContext {
+                    // recognize these typedefs by the name of the typedef type.
+                    let id_for_name = |name| -> Option<_> {
+                        let kind = match name {
+                            "intmax_t" => CTypeKind::IntMax,
+                            "uintmax_t" => CTypeKind::UIntMax,
+                            "intptr_t" => CTypeKind::IntPtr,
+                            "uintptr_t" => CTypeKind::UIntPtr,
+                            // unlike `size_t`, `ssize_t` does not have a clang-provided `#define`.
+                            "ssize_t" => CTypeKind::SSize,
+                            // macOS defines `ssize_t` from `__darwin_ssize_t` in many headers,
+                            // so we should handle `__darwin_ssize_t` itself.
+                            "__darwin_ssize_t" => CTypeKind::SSize,
+                            "__uint8_t" => CTypeKind::UInt8,
+                            "__uint16_t" => CTypeKind::UInt16,
+                            "__uint32_t" => CTypeKind::UInt32,
+                            "__uint64_t" => CTypeKind::UInt64,
+                            "__uint128_t" => CTypeKind::UInt128,
+                            "__int8_t" => CTypeKind::Int8,
+                            "__int16_t" => CTypeKind::Int16,
+                            "__int32_t" => CTypeKind::Int32,
+                            "__int64_t" => CTypeKind::Int64,
+                            "__int128_t" => CTypeKind::Int128,
+                            // macOS stdint.h typedefs [u]intN_t directly:
+                            "int8_t" => CTypeKind::Int8,
+                            "int16_t" => CTypeKind::Int16,
+                            "int32_t" => CTypeKind::Int32,
+                            "int64_t" => CTypeKind::Int64,
+                            "uint8_t" => CTypeKind::UInt8,
+                            "uint16_t" => CTypeKind::UInt16,
+                            "uint32_t" => CTypeKind::UInt32,
+                            "uint64_t" => CTypeKind::UInt64,
+                            _ => {
+                                log::debug!("Unknown fixed-size type typedef {name}!");
+                                return None;
+                            }
+                        };
+                        log::trace!("Selected kind {kind} for typedef {name}");
+                        Some(CQualTypeId::new(
+                            self.typed_context.type_for_kind(&kind).unwrap(),
+                        ))
+                    };
+                    let file = self
+                        .typed_context
+                        .files
+                        .get(self.typed_context.file_map[node.loc.fileid as usize]);
+                    let file = file.unwrap();
+                    if let Some(path) = file.path.as_ref() {
+                        if let Some(filename) = path.file_name() {
+                            if filename == "stdint.h"
```

Comment:
Do all of these headers come from clang, or are some from glibc? https://github.com/llvm/llvm-project/tree/main/clang/lib/Headers seems to have `stdint.h`, `stddef.h` and `inttypes.h`, but I'm not sure what gets installed.

Other C libraries structure these headers a bit differently. See for example musl's definitions at https://git.musl-libc.org/cgit/musl/tree/include/alltypes.h.in.

Reply:
At least some of these are expected to come from the C library, not Clang. It's likely that we might need additional handling to support musl and other BSDs. I'm not particularly happy with this unportable logic, but I'm not aware of a predicate that catches exactly what we want here (headers used in the implementation of standard types' definitions) in a portable way.